cmake/FortranEnvironment.cmake

```diff
@@ -0,0 +1,118 @@
+# ================================
+# CMAKE Script for setting up the fortran compiling and linking flags for different operating systems and compilers
+
+# Call this cmake file from your project's CMakeLists.txt using 'include(/path/to/FortranEnvironment.cmake)'
+
+# ++++++++++++++++++++++++++++++++
+
+
+enable_language(Fortran)
+
+option (BUILD_SHARED_LIBS "Shared or static libraries"  ON)
+
+# Check if linux
+if(UNIX AND NOT APPLE)
+  set(LINUX TRUE)
+endif()
+
+
+# ================================
+# Default the Build type to RELEASE
+# ================================
+# Make sure the build type is uppercase
+string(TOUPPER "${CMAKE_BUILD_TYPE}" BT)
+
+if(BT STREQUAL "RELEASE")
+  set(CMAKE_BUILD_TYPE RELEASE CACHE STRING
+    "Choose the type of build, options are DEBUG, or RELEASE."
+    FORCE)
+elseif(BT STREQUAL "DEBUG")
+  set(CMAKE_BUILD_TYPE DEBUG CACHE STRING
+    "Choose the type of build, options are DEBUG, or RELEASE."
+    FORCE)
+elseif(NOT BT)
+  set(CMAKE_BUILD_TYPE DEBUG CACHE STRING
+    "Choose the type of build, options are DEBUG, or RELEASE."
+    FORCE)
+  message(STATUS "CMAKE_BUILD_TYPE not given, defaulting to DEBUG.")
+else()
+  message(FATAL_ERROR "CMAKE_BUILD_TYPE not valid, choices are DEBUG, or RELEASE.")
+endif(BT STREQUAL "RELEASE")
+
+# ++++++++++++++++++++++++++++++++
+
+# ================================
+# Find the openmp package and add compiler-specific flags
+# ================================
+find_package(OpenMP)
+if(OPENMP_FOUND)
+  set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} ${OpenMP_Fortran_FLAGS}")
+  set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} ${OpenMP_EXE_LINKER_FLAGS}")
+endif()
+# ++++++++++++++++++++++++++++++++
+
+# ================================
+# Set gfortran compile flags
+# ================================
+message("\n")
+if(CMAKE_Fortran_COMPILER_ID MATCHES "GNU")
+  include(${CMAKE_MODULE_PATH}/gfortran_flags.cmake)
+
+# ================================
+# Set INTEL compile flags
+# ================================
+elseif(CMAKE_Fortran_COMPILER_ID MATCHES "Intel")
+  include(${CMAKE_MODULE_PATH}/intel_flags.cmake)
+
+elseif(CMAKE_Fortran_COMILER_ID MATCHES "PGI")
+  message(FATAL_ERROR "Need to define flags for PGI")
+
+endif()
+
+# ================================
+# Set the output directories for compiled libraries and module files.
+# Paths are relative to the build folder where you call cmake
+# ================================
+set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}/../lib) # Static library location
```

Comment:
To make it clear these lines are creating side effects (creating artifacts outside of the build directory) would it be possible to do like:
```cmake
set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${PROJECT_SOURCE_DIR}/lib)
```

Reply:
Very good point. I think, it is generally not a nice idea to create anything outside of the build folder (not even in the source folder), unless the user explicitly asks for it. So, why not something like
```
set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}/_install/lib)
```
and the user can override it, in case it should be put somewhere else.